pkg/query/querier.go

```diff
@@ -208,7 +208,10 @@ func (q *querier) Select(_ bool, hints *storage.SelectHints, ms ...*labels.Match
 	// The querier has a context but it gets canceled, as soon as query evaluation is completed, by the engine.
 	// We want to prevent this from happening for the async storea API calls we make while preserving tracing context.
 	ctx := tracing.CopyTraceContext(context.Background(), q.ctx)
```

Comment:
Does q.ctx has `ctx, cancel := context.WithTimeout(ctx, q.selectTimeout)`? select Timeout? I don't think we ever set this timeout into q.ctx

Reply:
Question: Ideally we want request to override timeout - let's see if we can do it.

Reply:
For q.ctx, 'per-request timeout' is set at  [Thanos Query API](https://github.com/thanos-io/thanos/blob/c634e312e080f531747412285262fea902033215/pkg/api/query/v1.go#L242) and later in call chain 'engine timeout' is set at [Prometheus layer](https://github.com/prometheus/prometheus/blob/41247b69507342b8d26110207af988ad5a1310e5/promql/engine.go#L440) . So the downstream call-chain at package like pkg/query, would not carry per-request timeout any more. 

Added the unit tests to test if lower value out of {global Query timeout, per-request timeout} is used to execute queries.